### PR TITLE
[8.0][PORT] BOM stock

### PR DESCRIPTION
**Pull request by SodexisTeam**

### CODE REVIEW

File: bom_stock/__init__.py [hunk 1]
```diff
@@ -18,6 +18,6 @@
 #
```
Comment (elicoidal) [inline bom_stock/__init__.py]: @SodexisTeam Thanks for your contribution to the OCA. 
Please follow the guidelines from:
https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
You need to create a README file following: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
Besides, for all your py files, please use short headers following:
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

File: bom_stock/__openerp__.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
 {
     'name': 'Compute Stock from BoM',
-    'version': '5.0.2',
+    'version': '8.0.0',
```
Comment (elicoidal) [inline bom_stock/__openerp__.py]: 8.0.1.0.0

File: bom_stock/__openerp__.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
...
     'category': 'Generic Modules/Others',
     'description':
```
Comment (elicoidal) [inline bom_stock/__openerp__.py]: remove (in the README file now). 
Adding screenshots in the README file is a must (nice to have)

File: bom_stock/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
```
Comment (elicoidal) [inline bom_stock/models/product_product.py]: use simple headers

File: bom_stock/models/product_product.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        bom_id = bom_obj._bom_find(product_id=self.id, properties=[])
+        if bom_id:
+            prod_min_quantities = []
+            bom = bom_obj.browse(bom_id)
+            if bom.bom_line_ids:
+                stop_compute_bom = False
+                for line in bom.bom_line_ids:
+                    prod_min_quantity = 0.0
+                    bom_qty = line.product_id[stock_field]
+                    line_product_qty = uom_obj._compute_qty_obj(line.product_uom, line.product_qty, line.product_id.uom_id,)
```
Comment (elicoidal) [inline bom_stock/models/product_product.py]: please follow PEP8 for all the code

**Comment by elicoidal** [comment]
@SodexisTeam For the CLA, please follow instructions here: https://odoo-community.org/page/cla
I did no code inspection, only formal review.
You need to follow the guidelines for headers, README, **openerp**.py and PEP8

**Comment by stephankeller** [comment]
@elicoidal thanks for helping with this, we are working on the changes.
I did sign an Entity CLA in the past.  I think all what is missing is the list of github accounts, where should we send that?

Thanks,
Stephan.

**Comment by atchuthan** [comment]
@elicoidal This module was created by CamptoCamp and we have ported this module to the new api of Odoo.  So, who should I add as author in **openerp**.py and in license headers?

**Comment by elicoidal** [comment]
@stephankeller the CLA in our database is linked to Sodexis but @atchuthan  is not declare under this ECLA.
Can you send an email at cla at odoo - community dot org with full name and email address so that we add him below Sodexis?

**Comment by elicoidal** [comment]
@atchuthan yes, you should add in the openerp.py and py file headers Camptocamp (check the copyright of original module to add it here.
Authors are companies and contributors individuals. Feel free to add significant contributors from them.

File: bom_stock/README.rst [hunk 6]
```diff
@@ -0,0 +1,59 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+==============
```
Comment (elicoidal) [inline bom_stock/README.rst]: lines should be until the end of the sentence

File: bom_stock/README.rst [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+Compute Product Stock from its Bill of Material
+==============
+
+Compute the BOM  stock Value. 
+BoM Stock Value are computed by:
+(`Reference stock` of `Product` + How much could I produce of that `Product` according to the component's `Reference Stock`)
```
Comment (elicoidal) [inline bom_stock/README.rst]: please try to maintain the 80 char. limit her as well

File: bom_stock/README.rst [hunk 8]
```diff
@@ -0,0 +1,59 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+==============
+Compute Product Stock from its Bill of Material
+==============
+
+Compute the BOM  stock Value.
```
Comment (elicoidal) [inline bom_stock/README.rst]: Simplify the 2 lines in one:
Compute the BOM  stock Value based on:

File: bom_stock/README.rst [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+Compute the BOM  stock Value. 
+BoM Stock Value are computed by:
+(`Reference stock` of `Product` + How much could I produce of that `Product` according to the component's `Reference Stock`)
+
+This reference stock can be chosen by company through a selection field
+and can be one of the available stock quantity computed in the system :
+Available stock, Virtual stock, immediately_usable stock (from stock_available_immediately).
+
+Known issues / Roadmap
```
Comment (elicoidal) [inline bom_stock/README.rst]: remove empty section

**Comment by elicoidal** [comment]
small details left.
other than that LGTM (no test nor technical review)

**Comment by stephankeller** [comment]
@elicoidal I sent the email.

For info here is the list of github users that needs to be added:
Atchuthan: atchuthan atchuthan@sodexis.com
Dhinesh: dvdhinesh dhinesh@sodexis.com
John Britto:: johnbritto1 john@sodexis.com
SodexisTeam: SodexisTeam dev@sodexis.com
Stephan Keller: stephankeller skeller@sodexis.com
Suganthi Karunanithi: suganthikarunanithi ksuganthi@sodexis.com
Xavier Dass: xavier-dass xavier@sodexis.com

Thanks,

**Comment by stephankeller** [comment]
@elicoidal  I got the email confirmation that the github accounts were received.

**Comment by max3903** [comment]
CLA fixed.

**Comment by stephankeller** [comment]
@max3903 Thanks!

**Comment by SodexisTeam** [comment]
@elicoidal It seems flake8 & pylint test are successful but not with Odoo/OCB test in travis.
Also, it seems cl/runbot test failed too...
Please clarify what I am missing?

**Comment by elicoidal** [comment]
cc @pedrobaeza

File: bom_stock/README.rst [hunk 10]
```diff
@@ -0,0 +1,53 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=================================================================
+Compute Product Stock from its Bill of Material
+=================================================================
```
Comment (pedrobaeza) [inline bom_stock/README.rst]: It must have the same length as the string

File: bom_stock/README.rst [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+Compute the BOM stock Value based on:
+(`Reference stock` of `Product` + How much could I produce of that 
+`Product` according to the component's `Reference Stock`)
+
+This reference stock can be chosen by company through a selection field
```
Comment (pedrobaeza) [inline bom_stock/README.rst:16]: Put this in usage section

Comment (pedrobaeza) [inline bom_stock/README.rst:16]: Reading twice, maybe better on Configuration

File: bom_stock/README.rst [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+and can be one of the available stock quantity computed in the system :
+Available stock, Virtual stock, immediately_usable stock 
+(from stock_available_immediately).
+
+Bug Tracker
```
Comment (pedrobaeza) [inline bom_stock/README.rst:35]: Add "Try me on runbot" button on Usage section

File: bom_stock/README.rst [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+Camptocamp SA
+Sodexis <dev@sodexis.com>
```
Comment (pedrobaeza) [inline bom_stock/README.rst:58]: Add the mentioned contributors

File: bom_stock/__openerp__.py [hunk 14]
```diff
@@ -1,49 +1,31 @@
...
-    'data': ['bom_stock_view.xml'],
-    'demo': [],
-    'test': ['tests/test_bom_stock.yml',
-             ],
-    'installable': False,
-    'active': False,
+    "name": "Compute Stock from BoM",
+    "summary": "Compute Product Stock from its Bill of Material",
+    "version": "8.0.1.0.0",
+    "category": "Uncategorized",
```
Comment (pedrobaeza) [inline bom_stock/__openerp__.py]: There should be a better category, like "Manufacturing"

File: bom_stock/views/product_template_view.xml [hunk 15]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- © 2015 Camptocamp SA, Sodexis
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+     
+<openerp>
+	<data>
```
Comment (pedrobaeza) [inline bom_stock/views/product_template_view.xml]: Don't indent with tabs, but spaces.

File: bom_stock/views/res_company_view.xml [hunk 16]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- © 2015 Camptocamp SA, Sodexis
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+
+<openerp>
+	<data>
```
Comment (pedrobaeza) [inline bom_stock/views/res_company_view.xml]: Don't indent with tabs, but spaces.

File: bom_stock/models/product_product.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    _inherit = 'product.product'
+
+    @api.multi
+    def _bom_stock_mapping(self):
+        return {'real': 'qty_available',
+                'virtual': 'virtual_available',
+                'immediately': 'immediately_usable_qty'}
+
+    @api.one
+    def _compute_bom_stock(self):
```
Comment (pedrobaeza) [inline bom_stock/models/product_product.py:19]: As this is not stored, the performance will suffer a lot. You should set correctly field dependencies, and make it stored.

**Comment by pedrobaeza** [comment]
Travis fails because this module is depending on stock_available_immediately, and that module is not here. You have to add to the oca_dependencies.txt file in your PR the repo where it's hosted. Is that dependency really needed?

**Comment by pedrobaeza** [comment]
Module name should be _mrp_bom_stock_, and I'm not sure that this is related to product-kitting at all.

File: bom_stock/README.rst (hunk 13 above)
Comment (atchuthan) [inline bom_stock/README.rst:58]: @pedrobaeza 
do you mean the contributors available in this github repo?

Users who have contributed to this file
    @gurneyalex gurneyalex
    @yvaucher yvaucher
    @guewen guewen
    @sbidoul sbidoul

Comment (pedrobaeza) [inline bom_stock/README.rst:58]: Yes

File: bom_stock/models/product_product.py (hunk 17 above)
Comment (atchuthan) [inline bom_stock/models/product_product.py:19]: `@api.depends(
        'product_tmpl_id.bom_ids',
        'product_tmpl_id.bom_ids.bom_line_ids',
        'product_tmpl_id.bom_ids.bom_line_ids.product_id',
        'product_tmpl_id.bom_ids.bom_line_ids.product_id.qty_available',
        'product_tmpl_id.bom_ids.bom_line_ids.product_id.virtual_available',
        'product_tmpl_id.bom_ids.bom_line_ids.product_id.immediately_usable_qty',
    )
`
def _compute_bom_stock(self):

BOM manipulation triggers the functional field computation but not the qty_available, virtual_available and immediately_usable_qty changes in the BOM products.

Am I missing some feature with the @api.depends or with the relation used in depends?

**Comment by stephankeller** [comment]
@pedrobaeza do you want us to change the name of the module?  I am not sure what the correct approach is here, that module was there in older version, so do we keep the same name or do we change it?  We will follow OCA decision on this.

We needed this for one of our customers, so we went ahead with the port.  If you don't want this in OCA anymore just let me know.  We are not advocating that is the right thing to do or not, we are just trying to contribute what we did.  I don't think we changed any of the functionalities.  If OCA decide to merge with the other module you mentioned, then we will try to find the time to help. 

@sebastienbeau @jgrandguillaume

**Comment by pedrobaeza** [comment]
I don't have a special preference of one module over the other, because I haven't studied any of them. I was just telling about make any convergence of both to avoid duplicates.

**Comment by atchuthan** [comment]
Travis test shows an error with lxml package import.
_ImportError: No module named lxml_

**Comment by pedrobaeza** [comment]
They have blacklisted python-lxml package. I have opened an issue for getting this again whitelisted: https://github.com/travis-ci/apt-package-whitelist/issues/2338

**Comment by atchuthan** [comment]
@pedrobaeza, Should I squash all the commits to a single commit?

**Comment by pedrobaeza** [comment]
You have to keep the relevant commits that means anything in your development cycle, not bugfixings nor PEP8 or minor changes that you have forgotten, but that's only my commit criteria. Others may want to be more exhaust on the commits.

**Comment by atchuthan** [comment]
@pedrobaeza, Label needs to be changed to "Needs review"

**Comment by gurneyalex** [comment]
:+1:

**Comment by jgrandguillaume** [comment]
Hi,

I understand may be this module should be in an other repo (to be discussed). But currently he's here, and ported, so go for it. For the name change, I think mrp_\* is not appropriate neither (nothing to do with manufacturing order). We just use the bom here...

I've tested it and it worked., so I'm :+1: 

Regards,

Joël

**Comment by gurneyalex** [comment]
How is this different from https://github.com/OCA/stock-logistics-warehouse/tree/8.0/stock_available_mrp ?

**Comment by OSguard** [comment]
it is the same function

i recommend:
- close this PR
- remove 8.0 and 9.0 branch of this repository as this is not needed any more

**Comment by atchuthan** [comment]
We created 2 PRs to remove this module from the repository for both 8.0(https://github.com/OCA/product-kitting/pull/12) and 9.0(https://github.com/OCA/product-kitting/pull/13) versions.

**Pull request closed without merging**